horovod/run/common/util/safe_shell_exec.py

```diff
@@ -142,26 +135,17 @@ def kill_executor_children_if_sigterm_received():
         stdout = sys.stdout
     if stderr is None:
         stderr = sys.stderr
-    stdout_fwd = threading.Thread(target=forward_stream, args=(stdout_r, stdout, 'stdout', index))
-    stderr_fwd = threading.Thread(target=forward_stream, args=(stderr_r, stderr, 'stderr', index))
-    stdout_fwd.start()
-    stderr_fwd.start()
 
-    def kill_middleman_if_master_thread_terminate():
-        event.wait()
-        try:
-            os.kill(middleman_pid, signal.SIGTERM)
-        except:
-            # The process has already been killed elsewhere
-            pass
+    stdout_fwd = in_thread(target=forward_stream, args=(stdout_r, stdout, 'stdout', index))
+    stderr_fwd = in_thread(target=forward_stream, args=(stderr_r, stderr, 'stderr', index))
 
     # TODO: Currently this requires explicitly declaration of the event and signal handler to set
     #  the event (gloo_run.py:_launch_jobs()). Need to figure out a generalized way to hide this behind
     #  interfaces.
+    stop = None
     if event is not None:
-        bg_thread = threading.Thread(target=kill_middleman_if_master_thread_terminate)
-        bg_thread.daemon = True
-        bg_thread.start()
+        stop = threading.Event()
+        on_event(event, os.kill, (middleman_pid, signal.SIGTERM), stop=stop)
```

Comment:
Should this set `silent=True` to be consistent with previous `kill_middleman_if_master_thread_terminate()` behavior?

Reply:
You are totally right, this one place was the reason I added it, but refactoring around must have dropped that argument here. Well spotted.